base/src/main/java/org/aya/tyck/pat/PatternTycker.java

```diff
@@ -0,0 +1,403 @@
+// Copyright (c) 2020-2022 Tesla (Yinsen) Zhang.
+// Use of this source code is governed by the MIT license that can be found in the LICENSE.md file.
+package org.aya.tyck.pat;
+
+import kala.collection.SeqView;
+import kala.collection.immutable.ImmutableSeq;
+import kala.collection.mutable.MutableList;
+import kala.control.Result;
+import kala.tuple.Tuple;
+import kala.tuple.Tuple3;
+import kala.value.MutableValue;
+import org.aya.concrete.Expr;
+import org.aya.concrete.Pattern;
+import org.aya.core.def.CtorDef;
+import org.aya.core.def.Def;
+import org.aya.core.pat.Pat;
+import org.aya.core.pat.PatMatcher;
+import org.aya.core.repr.AyaShape;
+import org.aya.core.term.*;
+import org.aya.core.visitor.DeltaExpander;
+import org.aya.core.visitor.Expander;
+import org.aya.core.visitor.Subst;
+import org.aya.generic.Constants;
+import org.aya.generic.util.InternalException;
+import org.aya.generic.util.NormalizeMode;
+import org.aya.pretty.doc.Doc;
+import org.aya.ref.AnyVar;
+import org.aya.ref.GenerateKind;
+import org.aya.ref.LocalVar;
+import org.aya.tyck.ExprTycker;
+import org.aya.tyck.error.TyckOrderError;
+import org.aya.tyck.trace.Trace;
+import org.aya.tyck.tycker.TyckState;
+import org.aya.util.Arg;
+import org.aya.util.reporter.Problem;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.jetbrains.annotations.UnknownNullability;
+
+/**
+ * A Pattern Tycker for only one use.
+ */
+public final class PatternTycker {
+  public final @NotNull ExprTycker exprTycker;
+  public final @NotNull TypedSubst bodySubst;
+  private final @NotNull TypedSubst sigSubst = new TypedSubst();
+  private @NotNull Def.Signature<?> signature;
+  private @NotNull SeqView<Arg<Pattern>> patterns;
+  private final @NotNull MutableList<Arg<Pat>> wellTyped = MutableList.create();
+  private Term.@UnknownNullability Param currParam = null;
```

Comment:
I am not sure about this: `private Term.@UnknownNullability Param currParam = null;`

Reply:
What specifically are you not sure about?

Reply:
Maybe we should pass `currParam` by argument? idk!

Reply:
A mutable field is fine. That's the point of OOP! :)